### PR TITLE
[processor/tailsampling] add OTTL Condition policy

### PR DESCRIPTION
**Pull request by jiekun**
**Description:** We could solve https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20294 by introducing OTTL to the tail sampling processor. By using the new policy, the processor can make sampling decision based on the OTTL statement.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20294

**Testing:** See `processor/tailsamplingprocessor/internal/sampling/ottl_test.go`.

**Documentation:** See `README.me`

### CODE REVIEW

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (TylerHelmuth) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:41]: I haven't done an in-depth look at this PR yet but checkout internal/filter/filterottl. I think it can replace a lot of this code.

File: processor/tailsamplingprocessor/config.go
Comment (TylerHelmuth) [inline processor/tailsamplingprocessor/config.go:219]: I strongly recommend you let users decide which ottl.ErrorMode is used with the statements. Checkout transformprocessor or filterprocessor for an example of that configuration.

Comment (jiekun) [reply processor/tailsamplingprocessor/config.go:219]: @TylerHelmuth Thanks for the early review! I **love** how our community pays attention to new contributions.

I've awarded how those configs should be decided (mostly comes from user input, like how we've done in transform processor / other internal pkg).

The current code is written like this due to the lack of understanding of OTTL-related packages. I am trying to implement and run it first. This should be well optimized, especially on how OTTL pkg is used, and how the code structure is organized.

Actually, the field you mentioned is an option for config in previous codes, like how we are doing in: `processor/transformprocessor/internal/common/traces.go`
```go
func (pc TraceParserCollection) ParseContextStatements(contextStatements ContextStatements) (consumer.Traces, error) {
                ...
		sStatements := ottlspan.NewStatements(parsedStatements, pc.settings, ottlspan.WithErrorMode(pc.errorMode))
		...
```

I removed them for a quick test. 

I think we will post a correct commit for review later and remove the WIP prefix before ping ya for review. And again your early suggestions are valuable for rookies like me. Please feel free to point out more issues ^_^

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (TylerHelmuth) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:48]: Can this function take a `component.TelemetrySettings` instead and pass in the processor's instance?

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (TylerHelmuth) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:87]: Since we have access to the skipExpr here via `osf` I think we can remove the need for `evalFuncChain`. Checkout how the filterprocessor handles span and spanevent conditions: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/af0f695997043ac28db2e2662ffd28fc799bb8cb/processor/filterprocessor/traces.go#L95-L115

**Review by TylerHelmuth — COMMENTED** [review]
Please update the README and add a changelog entry

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: OKie it's done inside the loop now. I feel a little bit uncomfortable when writing down so many for loops (lmao). Is it common when processing span-level stuff? It's identical to the filter processor before but I moved them outside the loop for to make it neater...

File: processor/tailsamplingprocessor/config.go
Comment (jiekun) [reply processor/tailsamplingprocessor/config.go]: Changed all statement(s) to condition(s).

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]:
```go
func NewOTTLStatementFilter(set component.TelemetrySettings, spanStatements, spanEventStatements []string, errMode ottl.ErrorMode) (PolicyEvaluator, error) {
```

Do you mean (currently) we should pass the logger by a `component.TelemetrySettings`? I am not so sure that the expected signature. But either way, this may need @jpkrohling 's opinion since all available filters are similar to the current format.

Comment (TylerHelmuth) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: The current implementation passes a noop logger to the filterottl function which isn't good. In general it's best to pass the component's telemetry settings to OTTL but if we can't do that at least using the logger in the new struct would be required. 

To us telemetry settings not all the other policies would have to change since this New function isn't part of the interface.

Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: I see. Thanks for pointing out my mistake. The logger is fixed already. Now I am thinking about how the signature should be and how to organize it in caller function .

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (TylerHelmuth) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: Yup, loops within loops is a very common pattern in processors when accessing data.  It's also slightly more efficient than your original approach which I believe looped through all the spans once for the Span Conditions and then looped through all the spans again for the Span Event conditions.  The filterprocessor method only loops through the spans 1 time.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (TylerHelmuth) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:84]: We can move this declaration down into the `if ocf.sampleSpanEventExpr != nil` block.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (TylerHelmuth) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:45]: This is better than before, but still not as good as using the component's instance to the telemetry settings.  Currently only the logger is used, but it is always possible OTTL will want something else in the struct in the future.  I believe the change would be to pass `component.TelemetrySettings` from `createTracesProcessor` -> `newTracesProcessor` -> `getPolicyEvaluator` -> `getSharedPolicyEvaluator`.  All the other NewPolicy functions would still get the Logger from that struct, which is how it currently works.

Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: OK, thanks for your timely review! so most functions accept a `logger`, which we've switched to a `settings`. Please review it again.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: Fixed as suggested.

**Comment by jiekun** [comment]
thanks @TylerHelmuth for triggering the CI workflow. May I ask if how could I handle the failed test? I think my PR is not relevent to `Integration_test_with_v3_configuration` test, tho

**Comment by atoulme** [comment]
Please rebase to latest main to pick up a fix.

**Comment by jiekun** [comment]
> Please rebase to the latest main to pick up a fix.

Thank you. Synced with upstream now.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: I believe this is expected but happy to hear more opinion from reviewers.

**Review by MovieStoreGuy — APPROVED** [review]
These changes appear sensible enough and have test coverage for them.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (TylerHelmuth) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:49]: We should use errorMode here instead of hard-coding propagate.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (TylerHelmuth) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:55]: We should use errorMode here instead of hard-coding propagate.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: Fixed.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: Fixed.

**Review by TylerHelmuth — APPROVED** [review]
This looks good from an OTTL perspective.  @jpkrohling something to consider for the future is that this policy option has overlapping capabilities with many other policy options.  IDK how much flexibility there is to change this processor's config, but it could be beneficial to deprecate the duplicate policies.

**Comment by MovieStoreGuy** [comment]
> This looks good from an OTTL perspective. @jpkrohling something to consider for the future is that this policy option has overlapping capabilities with many other policy options. IDK how much flexibility there is to change this processor's config, but it could be beneficial to deprecate the duplicate policies.

@TylerHelmuth , did you want create a follow up issue for this? Or should this be decided before this PR is merged?

**Comment by TylerHelmuth** [comment]
@MovieStoreGuy definitely doesn't need to hold up this PR. Only needs to be an issue if @jpkrohling wants to take the processor that direction.

**Comment by jiekun** [comment]
Alright I think I could solve the conflict if we are going to do the merge. So we don't need to trigger CI flow again and again.

File: processor/tailsamplingprocessor/and_helper.go
Comment (jpkrohling) [inline processor/tailsamplingprocessor/and_helper.go:23]: This could have been split from this PR, making it easier to review/merge.

File: processor/tailsamplingprocessor/config_test.go
Comment (jpkrohling) [inline processor/tailsamplingprocessor/config_test.go:125]: Out of curiosity: would single quotes work to wrap the attribute names (such as `test_attr_key_1`)?

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:34]: Any reason for this to be exported and not the others?

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:64]: Is there a way to make this message more useful? Like, by adding the trace ID to it? If the purpose is to give the user feedback that this works, you may want a metric counter instead.

**Comment by TylerHelmuth** [comment]
@jiekun I'd like the code owner @jpkrohling to review before we merge bc this is a significant change.  Feel free to hold off handling any merge conflicts until he has reviewed.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:96]: Same as the previous comment. While this is useful for developing this component, extracting value from this log line is hard when you are operating the collector. Perhaps a counter would make sense here as well?

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/ottl.go:108]: Same

File: processor/tailsamplingprocessor/internal/sampling/ottl_test.go
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/ottl_test.go:41]: Should this be an error condition? At least one of the conditions should be specified, I suppose.

File: processor/tailsamplingprocessor/internal/sampling/ottl_test.go
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/ottl_test.go:55]: I'm missing a case where `attributes["attr_k_1"] != "attr_v_1"` and the span attribute value would be `attr_v_2". This would yield `Sampled`, right?

File: processor/tailsamplingprocessor/internal/sampling/ottl_test.go
Comment (jpkrohling) [inline processor/tailsamplingprocessor/internal/sampling/ottl_test.go:72]: I think this matches the condition I was missing, but I would still like to have one "not matching" a specific value.

**Comment by jpkrohling** [comment]
I'm fine merging this as is if the items I mentioned during my review are addressed in a follow-up PR.

File: processor/tailsamplingprocessor/config_test.go
Comment (TylerHelmuth) [reply processor/tailsamplingprocessor/config_test.go:125]: No, OTTL requires double quotes `""`.  In situations like these you could create the strings using back ticks ` instead of ".

File: processor/tailsamplingprocessor/and_helper.go
Comment (TylerHelmuth) [reply processor/tailsamplingprocessor/and_helper.go:23]: True and it still could be broken out if you want

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (TylerHelmuth) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go:34]: Good catch, they shouldn't need to be exported.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go:64]: I prefer to add a `traceID` to it. 

Actually the metric counter would be better. However, we have this debug log for all other tail sampling policies. I think it's bad when people see the log from other policies but empty output for ottl. They may wonder if it's not working.

And it's under `Debug` log level which may help a very little bit. I think we should either replace all of them with a metric, or keep the log and add more information to make it useful. WDYT?

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/be4b4efbb2dd32c9b1c7ded8720979fdeac33cd2/processor/tailsamplingprocessor/internal/sampling/span_count_sampler.go#L41

File: processor/tailsamplingprocessor/internal/sampling/ottl_test.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl_test.go:41]: I referred to the impl of filterprocessor: 
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/be4b4efbb2dd32c9b1c7ded8720979fdeac33cd2/processor/filterprocessor/traces.go#LL45C3-L45C3

I think it's still good to stop user from using a blank condition(mistakenly). But I am not confident about the necessity to change it since we have validation at the begining of  `Evaluate`:
```go
	if ocf.sampleSpanExpr == nil && ocf.sampleSpanEventExpr == nil {
		return NotSampled, nil
	}
```

Feel free to let me know and I am ready to change the init function to return an `error` when empty ottl condition is set.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go:34]: Thank you. It's fixed to `errorMode` now.

File: processor/tailsamplingprocessor/and_helper.go
Comment (jiekun) [reply processor/tailsamplingprocessor/and_helper.go:23]: +1 OK to split. I will keep in mind that we should do exactly one thing in a single PR. 

Since we got many reviews in this PR already, reserving this change might save some time and efforts of our reviewer?

File: processor/tailsamplingprocessor/internal/sampling/ottl_test.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl_test.go:55]: That's correct. New case added as well.

File: processor/tailsamplingprocessor/internal/sampling/ottl_test.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl_test.go:79]: New case added.

File: processor/tailsamplingprocessor/internal/sampling/ottl.go
Comment (jiekun) [reply processor/tailsamplingprocessor/internal/sampling/ottl.go]: I removed 2 unnecessary logging and keep the entry one only.

**Comment by jiekun** [comment]
/ptal @jpkrohling Thanks for reviewing. I am lookging forward for more input since it's my first PR.

File: processor/tailsamplingprocessor/and_helper.go
Comment (TylerHelmuth) [reply processor/tailsamplingprocessor/and_helper.go:23]: Ya we'll leave it.

**Review by jpkrohling — APPROVED** [review]

**Review by TylerHelmuth — COMMENTED** [review]
@jiekun please resolve the conflicts then we'll be good to go.

**Comment by jiekun** [comment]
/ptal @TylerHelmuth. Thanks for reviewing. I love how the OTel community responds to newbies' PR. I'm really looking forward to my 2nd contribution.

**Pull request merged**
